### ISSUE TITLE
build.gradle -- add note for MyOpenWeatherMapApiKey

### ISSUE THREAD

**Issue by Cattalins** [udacity/Sunshine-Version-2#69]
I am a pretty new user to Android Studio, Gradle and everything else.
I have been having some trouble to get the project starting with the build.gradle file.

ON:

```
    buildTypes.each {
        it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', MyOpenWeatherMapApiKey
    }
```

It would be nice to have a note stating that your API Key should be inside single quotes AND double quotes.

LIKE:

```
    buildTypes.each {
        // Replace MyOpenWeatherMapApiKey with your API Key in the format '"abcdefgh012345"' 
        // Use both single quotes followed by double quotes.
        it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', MyOpenWeatherMapApiKey
    }
```

**Comment by jasjasbinks** [comment]
Thank you Cattalins, I also encountered this hiccup :)

**Comment by a1g0rithm** [comment]
I agree Cattalins.  Part of being a good coder is using comments to help clarify what the previous code means.

**Comment by morfioce** [comment]
That was starting to bother me and then I read this, so thank you Cattalins

**Comment by zmckenney** [comment]
Wow, thank you for this post!  This was something I also ran into that bothered me.  Thanks for the information!

**Comment by dragon9001** [comment]
I get stuck at the same situation. Thanks so much!!! cattallins

**Comment by DClayton12** [comment]
@Cattalins Good catch and Thank you! I was searching errors using gradle as the error was not obvious.

**Comment by asheshb** [comment]
Thanks a lot Cattalins. That was very helpful and time saving.

**Comment by JerryZaz** [comment]
On the bright side, getting this fixed was also a learning experience and YOU get to say that you did it without this post! You're awesome!

**Comment by Cattalins** [comment]
Yes, I know, but sadly when this get fixed, I will return to my anonymity.
Hahaha

**Comment by bavly** [comment]
thank you Cattalins very much this helped me

**Comment by nammshub** [comment]
Thank you so much for your help :)

**Comment by tkinetik1** [comment]
AMEN!

**Comment by dbwest** [comment]
that helped so much. Should've guessed to have done that when I saw the code it generated...

**Comment by thaibt** [comment]
Hi I have this problem with buildTypes.each

Error:(21, 0) Gradle DSL method not found: 'buildConfigField()'
Possible causes:<ul><li>The project 'Sunshine' may be using a version of Gradle that does not contain the method.
<a href="openGradleSettings">Gradle settings</a></li><li>The build file may be missing a Gradle plugin.
<a href="apply.gradle.plugin">Apply Gradle plugin</a></li>

And no, I am not missing a comma.

**Comment by shrenuj** [comment]
Earlier,I quit the tutorial coz android studio wont let me run the app coz of this gradle quotes issue.
Thanks a lot!

**Comment by thaibt** [comment]
For some reason, my gradle built only works when I simulate through a virtual phone, not my galaxy.... Which was weird because it was compatible.

**Comment by peterqz** [comment]
Also this works:
 it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', "\"abcdefg0123456\""

**Comment by sachinrkeche** [comment]
Good Catch..

**Comment by javiexpo** [comment]
You can also use the gradle.properties file tu define the constant MyOpenWeatherMapApiKey like this:

MyOpenWeatherMapApiKey = "abcdefg0123456"

And the project will sync without problem

**Comment by lucasassalti** [comment]
Thank u =)

**Comment by lishuopitt** [comment]
This link may be helpful http://stackoverflow.com/questions/33365650/cannot-resolve-symbol-c882c94be45fff9d16a1cf845fc16ec5

**Comment by ratulbside** [comment]
This helps me a lot .. :)

**Comment by javadKarbasian** [comment]
Thanks a lot

**Comment by Polyterative** [comment]
This helped me **a lot**, thanks!!!

**Comment by tusharbanne** [comment]
I should have seen this before figuring out what the problem is. It would have saved a lot of time.

**Comment by ghost** [comment]
Thank you so much for this post, i have spent last 2 hrs fixing this issue.

**Comment by ghost** [comment]
Thanks a lot Cattalins

**Comment by ghost** [comment]
buildTypes.each {
     it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', "\"c882c94be45fff9d16a1cf845fc16ec5\""
 }
//me a funcionado!!!

**Comment by tal32123** [comment]
How does one use this code:

MyOpenWeatherMapApiKey = "abcdefg0123456"

it doesn't seem to work for me.

**Comment by ghost** [comment]
Hi,
Please add this line in your build.gradle file

buildTypes.each {
it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', "'abcdefg0123456'"
}

And after that refer this by BuildConfig.OPEN_WEATHER_MAP_API_KEY in your
ForecastFragment class.

Think it must work.

**Comment by tal32123** [comment]
Thanks for your response, what is the point of this code then in the gradle properties file?:

MyOpenWeatherMapApiKey = "abcdefg0123456"

**Comment by keysona** [comment]
Thanks!
I also met this problem!
I just add single quote...

**Comment by DroidPulkit** [comment]
Thank you so Much
It solved the problem

**Comment by Abhi0725** [comment]
Thanks

**Comment by vikitripathi** [comment]
it helped !

**Comment by BrunoCodex** [comment]
Thanks :100:

**Comment by cp-jones** [comment]
Thanks! Spent about an hour looking for this fix!

**Comment by Heleni** [comment]
Thanks a lot!

**Comment by baishali-ghosh** [comment]
Thanks a lot. Got stuck here, kept applying " " and it still wouldn't render correctly in the build config. You saved me a lot of time.

**Comment by technoplato** [comment]
There's going to be a change coming that will make this process much more clear! Sorry for the current confusion. I agree, single quotes surrounding double quotes is very confusing.

**Comment by alketola** [comment]
Thank you, you saved my day! Started pondering about this a fed days ago... 
- However, I consider this a very educative catch. ;-)

**Comment by sidvenu** [comment]
@tal32123 and others who are looking why defining `MyOpenWeatherMapApiKey` in `gradle.properties` doesn't work. The people at Udacity tells us to create `gradle.properties` in `.gradle` directory. But there is already in the normal project folder.`(C/Users/.../Sunshine/gradle.properties)`. Put `MyOpenWeatherMapApiKey="{YOURKEYHERE}"` there and replace `{YOURKEYHERE}` with your key. and then leave the `build.gradle` file as it is as they gave here at github. It will work

**Comment by technoplato** [comment]
@SiddharthVenu The reason why you shouldn't put your API key in your project's (`C/Users/.../Sunshine/gradle.properties`) is because gradle recommends that file being checked in to version control. If you have your API key in that file, you shouldn't check it into version control. If you put your API key in your GLOBAL gradle.properties (~/.gradle/gradle.properties), it will be recognized by your project's build.gradle, but won't be checked into version control, which is really the goal here.  

Does that make sense?

**Comment by sidvenu** [comment]
@halfjew22 But it does not recognize the variable if I put it in `~/.gradle`

**Comment by technoplato** [comment]
@SiddharthVenu Uh oh! Shoot me an email (mlustig@google.com) and I'll help you fix that!

**Comment by ahewlett** [comment]
I just encountered this, thank you for posting the solution.

**Comment by wgblackmon** [comment]
Can't fix this problem: 
Error:(21, 0) Gradle DSL method not found: 'buildConfigField()'
Possible causes:<ul><li>The project 'Sunshine-Version-2' may be using a version of Gradle that does not contain the method.
<a href="open.wrapper.file">Open Gradle wrapper file</a></li><li>The build file may be missing a Gradle plugin.
<a href="apply.gradle.plugin">Apply Gradle plugin</a></li>

Contents of app\build.gradle key entry:  buildTypes.each {
        it.buildConfigField 'String', "515c28e51f05ae59db9fb0b73b90eea0"
    }

or '"515c28e51f05ae59db9fb0b73b90eea0"' etc. 

I'm using the latest version of Android Studio 2.2.1 . This is the first day of my Udacity Android Nanodegree class and I'm new to Gradle. Any help will be GREATLY appreciated!

**Comment by poph2** [comment]
@Cattalins Your name has been engraved electronic stones of android history. You will not be forgotten.

**Comment by smsubham** [comment]
thanks for this solution , saved me hours of frustration.

**Comment by jmkuruvilla** [comment]
Thank you! That saved some headache.

**Comment by Josephaguele** [comment]
Why does it (OPEN_WEATHER_MAP_API_KEY)  have to be written like this?
And what is the effect of having to code in the build area? 

Please I need an answer to this questions

**Comment by AliHG** [comment]
@Josephaguele : naming OPEN_WEATHER_MAP_API_KEY like this will make no difference you can name it as you like but when you call it in your code  "FetchWeatherTask" class  `BuildConfig.OPEN_WEATHER_MAP_API_KEY;`  make sure to change the name here too

**Comment by AliHG** [comment]
@Cattalins i'm pretty sure this is not how it's done 

 `buildTypes.each {
        it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', MyOpenWeatherMapApiKey
    }`

and MyOpenWeatherMapApiKey is like calling a variable you declared in  gradle.propertie

`MyOpenWeatherMapApiKey="3194b70c58f9e0bc45dfd2279c1696c587"` 
 in gradle.properties file your .gradle folder in your home directory 

when adding your api key in your build.gradle file you risk when using github t

**Comment by Josephaguele** [comment]
I tried coding it like everyone has said, but it hasn't worked for me yet because when I try to run the sunshine app, it still shows the run time error. 
Is this not the correct  code?

buildTypes.each{
it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY',  "\"MyOpenWeatherMapApiKey\""
}

Please can anyone help me out with this. And please must we have a correct and working API key before the Sunshine app will work?

**Comment by AliHG** [comment]
@Josephaguele  this is the working code you add these lines exactly as it is in build.gradle file inside the android { } and this is my code 
`apply plugin: 'com.android.application'

android {
    compileSdkVersion 21
    buildToolsVersion "21.1.2"

```
defaultConfig {
    applicationId "com.example.android.sunshine.app"
    minSdkVersion 10
    targetSdkVersion 21
    versionCode 1
    versionName "1.0"
}
buildTypes {
    release {
        minifyEnabled false
        proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
    }
}

buildTypes.each {
    it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY' , MyOpenWeatherMapApiKey
}
```

}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    compile 'com.android.support:appcompat-v7:21.0.2'
}
`

you only add 

`buildTypes.each {
        it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY' , MyOpenWeatherMapApiKey
    }`
in the correct place inside the android { } scope and  don't change any thing else in this file and click  _sync now_

and you must have a working api Key for the sunshine app to work and this is my key " 3194b70c58f9e0bc45d2279c1696c587 " if you wish to use it 

and to must add your api key to the **gradle.properties**  file located in  [C:\Users\your username\.gradle](url) in windows and in your Home directory in mac  and here is the code in my **gradle.properties** file 

`org.gradle.daemon=true
MyOpenWeatherMapApiKey="3194b70c58f9e0bc45d2279c1696c587"`

if you need any help i'm here

**Comment by Josephaguele** [comment]
Thanks,
I really appreciate. The key you gave to me worked. Please can I just use
it for the practical sessions. Please

**Comment by AliHG** [comment]
@Josephaguele it's yours :+1:

**Comment by Uiyama** [comment]
Thanks a lot. I really appreciate.

**Comment by Josephaguele** [comment]
Hi,

Please Is there any way I can get an updated explanation of Postal Code
Param. I really do not understand it .
I would be very grateful.
Thanks in anticipation.

**Comment by Josephaguele** [comment]
Please my android monitor screen is not showing the information as shown in the tutorial. 
When I click the refresh button, this is what shows up on my screen.
![capture](https://cloud.githubusercontent.com/assets/20628649/17641945/028be4b8-612c-11e6-8fda-0b162a93d1cb.JPG)
I hope this will do well.
The last two lines, which is the InputManager is what shows up on my android studio monitor screen when I click the refresh button

**Comment by kavoos** [comment]
If you create (global) **gradle.properties** yourself, you should make a **PROPERTIES FILE** and not TEXT DOCUMENT (via notepad or something like that!)
Your gradle.properties should be like this:
![after](https://cloud.githubusercontent.com/assets/6151245/18748615/61ba2c08-80d3-11e6-95a5-356981b8d309.PNG)
and it must be in **C:\Users[user name].gradle**
It contains: **MyOpenWeatherMapApiKey="xxxx1111xxxx"**
Your **build.gradle** should be like this: 
**...
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
    buildTypes.each {
        it.buildConfigField 'String', 'OPEN_WEATHER_MAP_API_KEY', MyOpenWeatherMapApiKey
    }
...**

**Comment by seiya021** [comment]
Thank's  Cattalins

**Comment by Josephaguele** [comment]
Thank you

**Comment by omerboyraz** [comment]
Thank you for help.

**Comment by rag-lab** [comment]
ok, when changing the build.gradle file double check you inserted the code in the correct one. there's 2 here (Project and app), the code should be on the build.grade (app)

**Comment by sachinrkeche** [comment]
Thanks ..

**Comment by AMT16** [comment]
Thank you , that was helpful
